File: code/raytracer/javascript/camera.js
/**
 * Camera functions
 */

//Returns the left,right,top,bottom bounds of the near plane in camera space
function setCameraBounds(camera, width, height)
{
	var bounds = new Object();
	bounds.t = (Math.abs(camera.near) * Math.tan(((camera.fov / 2.0) / 180.0) * Math.PI));
	bounds.b = -bounds.t;
	bounds.r = bounds.t * width / height;
	bounds.l = -bounds.r;
	camera.cameraBounds = bounds;
}

//Transforms pixel coordinate from image space to camera space
function pixelToCameraCoords(camera, i, j, width, height)
{
	var bounds = camera.cameraBounds;
	var u = bounds.l + (bounds.r - bounds.l) * (i + 0.5) / width;
	var v = bounds.b + (bounds.t - bounds.b) * (j + 0.5) / height;
	var w = -camera.near;
	return [u,v,w];
}

//Gets camera space coordinate basis
function setCameraCoordinatesBasis(camera)
{
    var w = normalize(sub(camera.position, camera.target));
    var u = normalize(cross(camera.up, w));
    var v = normalize(cross(w, u));
    camera.coordinateBasis = [u,v,w]
}

//Convert vector in camera space (cameraCoords) into world space for camera
function cameraToWorldCoords(camera, cameraCoords)
{
	var basis = camera.coordinateBasis;
	var worldCoords = camera.position;
	for(var i=0; i<3; i++)
		worldCoords = add(worldCoords, mult_scalar(cameraCoords[i], basis[i]));
	return worldCoords;
}

File: code/raytracer/python/camera.py
from math import *
from collections import namedtuple

from linearAlgebra import *
#
# Camera defs
#
CameraBounds = namedtuple('CameraBounds', ['t', 'b', 'r', 'l'])


class Camera(object):
    def __init__(self, fov, position, up, target, near=0.1):
        self.fov = fov
        self.position = position
        self.up = up
        self.target = target
        self.near = near
        self.coordinateBasis = self.getCameraCoordinatesBasis()
        self.cameraBounds = None

    # Returns the left,right,top,bottom bounds of the near plane in camera space
    def setCameraBounds(self, width, height):
        t = (abs(self.near) * tan(((self.fov / 2.0) / 180.0) * pi))
        b = -t
        r = t * width / height
        l = -r
        self.cameraBounds = CameraBounds(t, b, r, l)

    # Transforms pixel coordinate from image space to camera space
    def pixelToCameraCoords(self, i, j, width, height):
        bounds = self.cameraBounds
        u = bounds.l + (bounds.r - bounds.l) * (i + 0.5) / width
        v = bounds.b + (bounds.t - bounds.b) * (j + 0.5) / height
        w = -self.near
        return [u, v, w]

    # Gets camera space coordinate basis
    def getCameraCoordinatesBasis(self):
        w = normalize(sub(self.position, self.target))
        u = normalize(cross(self.up, w))
        v = normalize(cross(w, u))
        return [u, v, w]

    # Convert vector in camera space (cameraCoords) into world space for camera
    def cameraToWorldCoords(self, cameraCoords):
        worldCoords = self.position
        for i in range(3):
            worldCoords = add(worldCoords, mult_scalar(cameraCoords[i], self.coordinateBasis[i]))
        return worldCoords

